vendor/noise_handshake/src/handshake.rs

```diff
@@ -0,0 +1,217 @@
+use futures::*;
+use snow::Session::{self, Handshake};
+use tokio::io::{AsyncRead, AsyncWrite};
+
+const NOISE_MAX_SIZE: usize = 65535;
+
+#[allow(missing_debug_implementations)]
+pub enum Step {
+    Read {
+        enc_buffer: [u8; NOISE_MAX_SIZE],
+        len: usize,
+    },
+    Write {
+        buffer: [u8; NOISE_MAX_SIZE],
+        written_bytes: usize,
+        len: usize,
+    },
+}
+
+impl Step {
+    fn read() -> Self {
+        Step::Read {
+            enc_buffer: [0u8; NOISE_MAX_SIZE],
+            len: 0,
+        }
+    }
+
+    fn write() -> Self {
+        Step::Write {
+            buffer: [0u8; NOISE_MAX_SIZE],
+            written_bytes: 0,
+            len: 0,
+        }
+    }
+}
+
+pub trait InitHandshake {
+    fn init_hand_shake<IO>(self, io: IO) -> NoiseHandshake<IO>
+    where
+        IO: AsyncRead + AsyncWrite;
+}
+
+impl InitHandshake for Session {
+    fn init_hand_shake<IO>(self, io: IO) -> NoiseHandshake<IO>
+    where
+        IO: AsyncRead + AsyncWrite,
+    {
+        match self {
+            Handshake(ref handshake_state) => NoiseHandshake {
+                next: if handshake_state.is_initiator() {
+                    Step::write()
+                } else {
+                    Step::read()
+                },
+                noise: Some(self),
+                io: Some(io),
+            },
+            _ => panic!(
+                "Noise Session in incorrect state, you should init before starting handshake"
+            ),
+        }
+    }
+}
+
+#[allow(missing_debug_implementations)]
+pub struct NoiseHandshake<IO: AsyncRead + AsyncWrite> {
+    next: Step,
+    noise: Option<Session>,
+    io: Option<IO>,
+}
+
+impl<IO: AsyncRead + AsyncWrite> NoiseHandshake<IO> {
+    fn wrap_up(&mut self) -> (Session, IO) {
+        let noise = self.noise.take().expect("We know it's a Some");
+        let noise = noise
+            .into_transport_mode()
+            .expect("Cannot go into transport mode despite handshake being finished");
+        let io = self.io.take().expect("We know it's a Some");
+        (noise, io)
+    }
+}
+
+impl<IO: AsyncRead + AsyncWrite> Future for NoiseHandshake<IO> {
+    type Item = (Session, IO);
+    type Error = std::io::Error;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        match self {
+            Self {
+                noise: Some(ref mut noise),
+                io: Some(_),
+                next:
+                    Step::Write {
+                        ref mut buffer,
+                        ref mut len,
+                        ..
+                    },
+            } if *len == 0 => {
```

Comment:
❌  Isn't len only 0 after you create a new `Step::Write`. Why not just write the message and get the len at the point you do `Step::write()` to avoid having three match arms.

Reply:
It's on purpose to make it clearer. Otherwise you end-up doing a lot in the same match arm.

Reply:
Hmm I think removing a match arm and bloating the other one would actually be clearer. You don't do any IO in that match arm so it doesn't feel like the others.

Reply:
I made a PR demonstrating how you can do this in two match arms #738

Reply:
Ok I did not understand it this way.

Reply:
done